File: pset6/readability.py
from cs50 import get_string
from sys import exit


def main():
    text = (get_string("Ingrese texto: "))

    # Create a list to store the results
    result = [0, 1, 0]

    # Parse the input text
    parse(text, result)

    # Process the results
    grade = compute(result)

    # Inform grade
    if (grade >= 16):
        print("Grade 16+")
    elif (grade < 1):
        print("Before Grade 1")
    else:
        print(f"Grade {grade}")
    exit(0)


def parse(txt, res):
    """
    Parse the received string. Sum every letter, word and sentence.
    """
    for char in txt:
        if (char.isalpha()):
            res[0] += 1
        if (char.isspace()):
            res[1] += 1
        if char in (".", "!", "?"):
            res[2] += 1
    return


def compute(res):
    """
    Receives a list of lettters, words and sencenteces
    and applies Coleman-Liau formulae.
    """

    # Avg of letters per 100 words
    L = res[0] * 100 / res[1]
    # Avg of sentences per 100 words
    S = res[2] * 100 / res[1]
    # Coleman-Liau formulae
    grade = 0.0588 * L - 0.296 * S - 15.8
    # Round the grade to the closest integer
    return round(grade)


main()

File: pset2/readability.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <ctype.h>
#include <cs50.h>

int parse(string);


int main(void)
{

    string input = get_string("Text:");

    int grade = parse(input);


    if (grade > 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }

    return 0;
}

// Parse the text and return a round number based on the given formula
int parse(string str)
{

    int i = 0;          // input's index

    int letters = 0;    //     #
    int words = 1;      // Counters
    int sentences = 0;  //     #

    while (str[i] != '\0')    // While not EOF
    {
        if ((str[i] >= 'a' && str[i] <= 'z') || (str[i] >= 'A' && str[i] <= 'Z'))
        {
            letters++;
        }
        else if (str[i] == ' ')    // End of word
        {
            words++;
        }
        else if (str[i] == '!' || str[i] == '?' || str[i] == '.')    // End of sentence
        {
            sentences++;
        }

        i++;
    }

    float L = (float) letters * 100 / (float) words;

    float S = (float) sentences * 100 / (float) words;

    float index = 0.0588 * L - 0.296 * S - 15.8;

    return round(index);
}